### DISCUSSION TITLE
Is a top level `await` equivalent possible?

### DISCUSSION THREAD

**Discussion by PEZ** [BetterThanTomorrow/joyride#58] [General]
In the context of a Joyride script, it would be great to be able to top-level define something that needs a promise to resolve. I guess I am looking for the equivalent of: 

```typescript
const something = await fetchSomething();

... script continues and can use `something`
```

Or even just firing of something for its side effects and waiting for it:

```typescript
await doThings();

... script continues when things are done
```

Is this even possible to get Joyride to support?

**Comment by borkdude** [comment]
@PEZ It is possible, but when we'd need to adapt the nbb way. Nbb has `(nbb.core/await ...)` to do such a thing.

I'm currently developing a `sci.async` namespace which should support this use case.

**Reply by borkdude** [reply]
> we'd need to adapt the nbb way

> Meaning to do everything async?

Yes. Every top level form has to be executed asynchronously. But sci.async will take care of this. The result of `sci.async/eval-string*` will return a promise, so that will also mean changes in joyride.

In general, a top level await should not be strictly necessary though, you can just compose `foo+` functions that return promises using promesa.

**Reply by PEZ** [reply]
> In general, a top level await should not be strictly necessary though, you can just compose foo+ functions that return promises using promesa.

You mean I can do this already today by instead of `def` using `defn` and to wrap the execution of the script in some kind of `main` function?

**Reply by borkdude** [reply]
yes. `def` also works, but the result is always a promise. you cannot escape that without any special means (like nbb does).

**Comment by mauricioszabo** [comment]
It is possible to have the "nbb way" that Borkdude already explained, and actually I wrote a code some time ago that patched SCI in some weird way that allowed to `deref` promises anywhere.

The only trouble is that it made SCI about 2x slower 🤣
